docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+> the last message to *FS*, the peer and *FS* run concurrently and
+> independently. There is no assumption on the rate at which a peer can
+> add blocks (e.g., it might be in the process of catching up itself).
+
+>  (*TD* might depend on timeouts etc. We suggest that an acceptable
+>  value for *TD* is in the range of apporx. 10 sec., that is the
+>  intervall between two calls `QueryStatus()`; see below.
+
+
+> Under [FS-ALL-CORR-PEER], if *peerIDs* contains a full node that is
+> "synchronized with the blockchain", and *blockchainheight* is the height
+> of the blockchain at time *term*, then  *terminationHeight* may even
+> achieve
+> *blockchainheight - TD / ETIME*;
+> cf. [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]. 
+
+
+
+
+
+#### Liveness
+
+<!--
+> liveness specifications in English. Possibly with timing/fairness requirements:
+e.g., if the component is connected to a correct full node and communication is
+reliable and timely, then something good happens eventually. 
+---->
+
+#### **[FS-VC-ALL-CORR-TERM]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* eventually terminates successfully.
+
+
+<!--
+> How is the problem statement linked to the "Sequential Problem statement". 
+Simulation, implementation, etc. relations 
+---->
+
+
+> We observe that all specifications that impose successful
+> termination at an acceptable height are all conditional under
+> [FS-ALL-CORR-PEER], that is, large parts of the current
+> implementations of FastSync (V2) are not fault-tolerant. We will
+> discuss this, and suggestions how to solve this after the
+> description of the current protocol.
+
+## Definitions
+
+<!--
+> In this section we become more concrete, with basic (abstracted) data types 
+---->
+<!--
+> some math that allows to write specifications and pseudo code solution below.
+Some variables, etc. 
+---->
+
+> We now introduce variables and auxiliary functions used by the protocol.
+
+#### Inputs
+- *startBlock*: the block Fastsync starts from
+- *startState*: application state corresponding to *startBlock.Height*
+
+#### **[FS-A-V2-INIT]**:
+- *startBlock* is from the blockchain
+- *startState* is the application state of the blockchain at Height *startBlock.Height*.
+
+#### Variables
+- *height*: initially *startBlock.Height + 1*
+  > height should be thought of the "height of the next block we need to download"
+- *state*: initially *startState*
+- *peerIDs*: peer addresses [FS-A-PEER-IDS](#fs-a-peer-ids)
+- *peerHeights*: stores for each peer the height it reported. initially 0
+- *pendingBlocks*: stores for each height which peer was
+  queried. initially nil for each height
+- *receivedBlocks*: stores for each height which peer returned
+  it. initially nil
+- *blockstore*: stores for each height greater than
+    *startBlock.Height*, the block of that height. initially nil for
+    all heights
+- *peerTimeStamp*: stores for each peer the last time a block was received
```

Comment:
There is also pendingTime variable in the code where time of the block request is stored. It is used as part of tryPrune function but only for the peer that is responsible for current height.

Reply:
I added at this point

- *pendingTime*: stores for each peer the last time a block was requested

and later, I updated:

#### **[FS-V2-PEER-REMOVE]**:
Periodically, *peerTimeStamp* and *peerRate* and *pendingTime* are
analyzed. 
If a peer *p*
has not provided a block recently (check of *peerTimeStamp[p]*) or it
has not provided sufficiently many data (check of *peerRate[p]*), then
*p* is removed from *peerIDs*. In addition, *pendingTime* is used to
estimate whether the peer that is responsible for current height is
still connected.